File: go/array/36_is_valid_sudoku.go
package array

// isValidSudoku 9 x 9 的数独是否有效
func isValidSudoku(board [][]byte) bool {

	// 记录元素e 出现在数组的哪一行，哪一列
	var rows, cols [9][9]int

	var subboxs [3][3][9]int

	for i, row := range board {
		for j, c := range row {
			// byte 类型 二维数组，里面是字符
			if c == '.' {
				continue
			}
			// 字符的内容是1-9， 用长度9 的数组存储，index 减一
			index := c - '1'

			// 例如如果为坐标(1,1)为8， 则在rows[i][8-1]的位置 +1
			rows[i][index]++
			cols[j][index]++
			subboxs[i/3][j/3][index]++

			if rows[i][index] > 1 || cols[j][index] > 1 || subboxs[i/3][j/3][index] > 1 {
				return false
			}

		}
	}

	return true
}

File: rust/array/36_is_valid_sudoku.rs


pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {

    let mut rows = vec![vec![0;9];9];
    let mut cols = vec![vec![0;9];9];

    // 注意此处理解为 三个 3 * 9 的二维数组
    let mut subboxs = vec![vec![vec![0;9];3];3];

    for (i, row) in board.iter().enumerate() {
        for (j, &c) in row.iter().enumerate() {
            if c == '.' {
                continue;
            }

            let index = (c as usize - '1' as usize);

            rows[i][index] += 1;
            cols[j][index] += 1;
            subboxs[i/3][j/3][index] +=1;

            if rows[i][index] > 1 || cols[j][index] > 1 || subboxs[i/3][j/3][index] > 1 {
                return false
            }
        }
    }

    true
}